### PR TITLE
Handle ordered/unordered lists

### PR DESCRIPTION
**Pull request by aringenbach**

### CODE REVIEW

File: crates/wysiwyg/src/composer_model.rs [hunk 1]
```diff
@@ -238,6 +242,47 @@ where
...
+                    let text = t.data();
+                    let list_node = DomNode::new_list(
+                        list_tag.to_html(),
+                        vec![DomNode::Container(ContainerNode::new_list_item(
+                            "li".to_html(),
+                            vec![DomNode::Text(TextNode::from(text.to_vec()))],
+                        ))],
+                    );
+                    self.state.dom.replace(range.node_handle, vec![list_node]);
+                }
```
Comment (andybalaam) [inline crates/wysiwyg/src/composer_model.rs:265]: We should panic in an else clause here - if it's not a text node, something is wrong.

File: crates/wysiwyg/src/composer_model.rs [hunk 2]
```diff
@@ -354,6 +436,61 @@ where
...
     }
+
+    fn add_list_item(&mut self, handle: DomHandle, location: usize) {
+        let list_node = self.state.dom.lookup_node_mut(handle);
+        if let DomNode::Container(list) = list_node {
+            list.append(DomNode::new_list_item(
+                "li".to_html(),
+                vec![DomNode::Text(TextNode::from("\u{200B}".to_html()))],
+            ));
+            self.state.start = Location::from(location + 1);
```
Comment (andybalaam) [inline crates/wysiwyg/src/composer_model.rs:447]: I think it would be good to select the hidden space, so should `self.state.start` be set to `location` instead of `location + 1`?

File: crates/wysiwyg/src/composer_model.rs [hunk 3]
```diff
@@ -354,6 +436,61 @@ where
...
+                    parent.remove(handle.index_in_parent());
+                    if parent.children().len() == 0 {
+                        parent.append(DomNode::Text(TextNode::from(
+                            "".to_html(),
+                        )));
+                    }
+                    self.state.start = Location::from(location);
+                    self.state.end = Location::from(location);
+                } else {
+                    panic!("List has no parent container")
```
Comment (andybalaam) [inline crates/wysiwyg/src/composer_model.rs:475]: Longer term, we should handle `li` not inside `ol` or `ul` - this could happen in HTML, even though we won't generate it. Probably fine for it to be a TODO for now, but we should think about it.

File: crates/wysiwyg/src/composer_model.rs [hunk 4]
```diff
@@ -354,6 +436,61 @@ where
...
+                }
+            } else {
+                list.remove(list_item_index);
+                let parent_handle = handle.parent_handle();
+                let parent_node = self.state.dom.lookup_node_mut(parent_handle);
+                if let DomNode::Container(parent) = parent_node {
+                    // TODO: should probably append a paragraph instead
```
Comment (andybalaam) [inline crates/wysiwyg/src/composer_model.rs:482]: Why a paragraph instead of a zero width space?

File: crates/wysiwyg/src/dom/dom_struct.rs [hunk 5]
```diff
@@ -98,6 +98,35 @@ where
...
 
+    pub fn find_parent_list_item(
+        &mut self,
+        child_handle: DomHandle,
+    ) -> Option<DomHandle> {
+        fn find_list_item<C: Clone>(
+            dom: &Dom<C>,
+            node: &DomNode<C>,
+        ) -> Option<DomHandle> {
+            if node.handle().has_parent() == false {
```
Comment (andybalaam) [inline crates/wysiwyg/src/dom/dom_struct.rs:109]: `            if !node.handle().has_parent() {`

File: crates/wysiwyg/src/dom/nodes/container_node.rs [hunk 6]
```diff
@@ -149,6 +178,22 @@ impl ContainerNode<u16> {
             handle: DomHandle::new_unset(),
         }
     }
+
+    pub fn is_empty_list_item(&self) -> bool {
+        match self.kind {
+            ContainerNodeKind::ListItem() => {
+                let html = self.to_html();
```
Comment (andybalaam) [inline crates/wysiwyg/src/dom/nodes/container_node.rs:185]: I'm a bit uncomfortable with converting to html here - I think `to_html` might have side effects in the future.

But this can be a TODO to fix later if you like.

File: crates/wysiwyg/src/dom/nodes/container_node.rs [hunk 7]
```diff
@@ -149,6 +178,22 @@ impl ContainerNode<u16> {
...
+                if html == "<li></li>".to_html()
+                    || html == "<li>\u{200b}</li>".to_html()
```
Comment (andybalaam) [inline crates/wysiwyg/src/dom/nodes/container_node.rs:187]: This `if` could be `return html == "<li></li>".to_html() || html == "<li>\u{200b}</li>".to_html()`

File: crates/wysiwyg/src/tests/test_lists.rs [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+#[test]
+fn test_ordered_list() {
+    let mut model = cm("|");
+    model.create_ordered_list();
+    assert_eq!(tx(&model), "<ol><li>|</li></ol>");
+    replace_text(&mut model, "abcd");
+    assert_eq!(tx(&model), "<ol><li>abcd|</li></ol>");
+    model.enter();
+    assert_eq!(tx(&model), "<ol><li>abcd</li><li>\u{200b}|</li></ol>");
```
Comment (andybalaam) [inline crates/wysiwyg/src/tests/test_lists.rs:29]: If we selected the space here...

File: crates/wysiwyg/src/tests/test_lists.rs [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    replace_text(&mut model, "efgh");
+    assert_eq!(tx(&model), "<ol><li>abcd</li><li>\u{200b}efgh|</li></ol>");
```
Comment (andybalaam) [inline crates/wysiwyg/src/tests/test_lists.rs:31]: ...then it would be gone here, which would be nice

File: crates/wysiwyg/src/tests/test_lists.rs [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+#[test]
+fn test_unordered_list() {
```
Comment (andybalaam) [inline crates/wysiwyg/src/tests/test_lists.rs:35]: I suspect we need tests for where some other text already exists.

I think the cursor will be in the wrong place in that case.

I am happy for this code to be merged, and then we fix that case. There are going to be LOADS of corner cases!

**Review by andybalaam — APPROVED** [review]
Happy to merge now, but we will need tons more test cases for this, and (obviously) we are going to need to handle cross-node selections as soon as that work is merged.

File: crates/wysiwyg/src/composer_model.rs (hunk 2 above)
Comment (aringenbach) [reply crates/wysiwyg/src/composer_model.rs:447]: I got a slightly weird selection issue with this on iOS, while this could probably be fixed with a platform-specific workaround, there is also the issue where we move the cursor somewhere else then move back to this TextNode...
Rather than reselecting the hidden space, it feels like it could be better to have some mechanic to automatically remove the hidden space when inserting near it ? (or even keeping it should be fine)

File: crates/wysiwyg/src/composer_model.rs (hunk 4 above)
Comment (aringenbach) [reply crates/wysiwyg/src/composer_model.rs:482]: I mean a paragraph with the zero width space inside it, perhaps it's not relevant.

File: crates/wysiwyg/src/dom/nodes/container_node.rs (hunk 6 above)
Comment (aringenbach) [reply crates/wysiwyg/src/dom/nodes/container_node.rs:185]: Yeah actually what we need here is a helper that actually just writes out the plain text from all children, this shouldn't be too hard to achieve I suppose. 
Because if I have something like "<li><strong>\u{200b}</strong></li>" I would expect to clear the `<li>` as well.

File: crates/wysiwyg/src/composer_model.rs (hunk 2 above)
Comment (andybalaam) [reply crates/wysiwyg/src/composer_model.rs:449]: Yeah. I was also wondering about just deleting all zero-width spaces whenever we convert to HTML.

Fine to leave it for now. When we add testcases all this is going to work itself out.

File: crates/wysiwyg/src/composer_model.rs (hunk 4 above)
Comment (andybalaam) [reply crates/wysiwyg/src/composer_model.rs:485]: I'm not sure whether we want that or not ... fine to leave a TODO to think about it.

File: crates/wysiwyg/src/dom/nodes/container_node.rs (hunk 6 above)
Comment (andybalaam) [reply crates/wysiwyg/src/dom/nodes/container_node.rs:186]: Good point. That is another test case we can create as soon as we have multi-node ranges.

**Review by andybalaam — APPROVED** [review]
Great!
